v2src/flash/attn_bwd_fused.cc

```diff
@@ -0,0 +1,233 @@
+      AOTRITON_NS::cdiv<uint32_t>(max_seqlen_k, params.BLOCK_N) +
+      AOTRITON_NS::cdiv<uint32_t>(max_seqlen_q, params.BLOCK_N) * (num_head_q / num_head_k),
+      uint32_t(params.K->size(1)),
+      params.num_seqlens == 0 ? uint32_t(params.Q->size(0)) : params.num_seqlens,
+    };
+    // std::cerr << "bwd_kernel_dk_dv grid conf " << grid.x << " " << grid.y << " " << grid.z << std::endl;
+    return grid;
+  };
+  constexpr int kMinHeadDimCompiled = 16;
+  int head_size = q.size(3);
+  const auto& compiled_head_dims = BwdKernelFuseMetadata::get_BLOCK_DMODEL_choices();
+  int head_size_rounded = round_value(head_size, compiled_head_dims);
+  if (head_size_rounded < 0) {
+#if AOTRITON_VERBOSE
+    std::cerr << "Head dimension " << head_size << " unsupported. ";
+    if (compiled_head_dims.empty()) {
+      std::cerr << "No head dimension (BLOCK_DMODEL) compiled into the binary." << std::endl;
+    } else {
+      std::cerr << "Maximal dimesion compiled into the binary is "
+                << compiled_head_dims.back()
+                << std::endl;
+    }
+#endif
+    return hipErrorInvalidValue;
+  }
+  int bias_type = 0;
+  if (b) {
+    bias_type = 1;
+  }
+  BwdKernelFuseParams params = {
+    .Q = &q,
+    .K = &k,
+    .V = &v,
+    .B = &b,
+    .Out = &out,
+    .DO = &dout,
+    .DK = &dk,
+    .DV = &dv,
+    .DQ = &dq,
+    .DB = &db,
+    .sm_scale = sm_scale,
+    .L = &softmax_lse,
+    .num_head_q = num_head_q,
+    .num_head_k = num_head_k,
+    .head_dim = head_size,
+    .cu_seqlens_q = &cu_seqlens_q,
+    .cu_seqlens_k = &cu_seqlens_k,
+    .num_seqlens = num_seqlens,
+    .max_seqlen_q = max_seqlen_q,
+    .max_seqlen_k = max_seqlen_k,
+    .dropout_p = dropout_p,
+    .philox_seed_ptr = &philox_seed,
+    .philox_offset1 = &philox_offset1,
+    .philox_offset2 = static_cast<uint64_t>(philox_offset2),
+    .BLOCK_DMODEL = head_size_rounded,
+    .CAUSAL = is_causal,
+    .ENABLE_DROPOUT = dropout_p > 0.0,
+    .PADDED_HEAD = head_size_rounded != head_size,
+    .BIAS_TYPE = bias_type,
+  };
+#if AOTRITON_BUILD_FOR_TUNING
+  if (extargs) {
+    params._has_preferred_kernel = extargs->fuse.force_kernel_index;
+    if (params._has_preferred_kernel == CppTuneSpecialKernelIndex::kSkipGPUCall) {
+        // std::cerr << "extargs->fuse.force_kernel_index = " << extargs->fuse.force_kernel_index << " EKI" << std::endl;
+        return hipSuccess;
+    }
+  }
+#endif
+  BwdKernelFuseContext context;
+  context.grid_calculator = grid_calculator;
+  err = context.lookup_optimal(params, arch);
+#if AOTRITON_BUILD_FOR_TUNING
+  if (extargs) {
+    extargs->fuse.total_number_of_kernels = params._total_number_of_kernels;
+    extargs->fuse.selected_kernel_psels = params._preferred_kernel_psels;
+    extargs->fuse.selected_kernel_copts = params._preferred_kernel_copts;
+  }
+#endif
+  if (err != hipSuccess) {
+    return err;
+  }
+  err = context.launch(params, stream);
+  return err;
+}
+
+hipError_t
+_attn_bwd_fused(T4 q,
```

Comment:
Why there are two APIs.
The bias tensor should not be ignored. User may pass nulltensor instead.